In [1]:
# ===== 기본 세트 (거의 항상 필요) =====
import pandas as pd                    # 데이터프레임 다루기 (표 형태 데이터 처리)
import numpy as np                     # 수치 계산, 배열 연산
import matplotlib.pyplot as plt        # 기본 시각화
import matplotlib.font_manager as fm   # 그래프에 한글 폰트 적용 시 필요
import seaborn as sns                  # matplotlib 기반의, 더 예쁘고 간결한 통계 시각화

# ===== 한글 폰트 설정 =====
font_path = "C:/Windows/Fonts/malgun.ttf"
fm.fontManager.addfont(font_path)
plt.rcParams["font.family"] = fm.FontProperties(fname=font_path).get_name()
plt.rcParams["axes.unicode_minus"] = False

# ===== 파일/경로/시스템 =====
import os                              # 파일 경로 확인, 파일 존재 여부(os.path.exists) 등
from pathlib import Path               # 폴더/파일 경로를 객체로 다루기 (work_dir / "파일명")
import joblib                          # 학습된 scaler/encoder를 파일로 저장하고 재사용
import platform                        # OS 종류 확인 (Windows/Mac/Linux 구분 필요할 때)
import warnings                        # 경고 메시지 제어(숨기기/표시)

# ===== 벤치마크/성능 측정 =====
import time                            # 코드 실행 시간 측정 (time.perf_counter())
import psutil                          # 메모리 사용량 측정 (rss_mb 등)
import gc                              # 메모리 강제 정리(garbage collection)

# ===== 결측치 탐색 시각화 =====
import missingno as msno               # 결측치 위치(matrix), 관계(heatmap) 시각화

# ===== 인터랙티브 시각화 =====
import plotly.express as px            # 마우스 hover/zoom 가능한 인터랙티브 그래프
import plotly.io as pio                # plotly 렌더러(출력 방식) 설정

# ===== 대용량 데이터 처리 =====
import polars as pl                    # pandas보다 빠른 대용량 데이터 처리 (Rust 기반)

# ===== 통계 검증 =====
import statsmodels.api as sm           # 회귀분석, 다중공선성(Cond. No.) 등 통계 검증
import statsmodels.formula.api as smf  # R 스타일 수식("y ~ x1 + x2")으로 회귀분석
from scipy import stats                # t-검정·카이제곱·정규성 검정 등 통계 검정 모음
from statsmodels.stats.power import TTestIndPower   # 검정력 분석(필요 표본수, 검정력 계산)
from sklearn.neighbors import NearestNeighbors      # 최근접 이웃 탐색 (경향점수 매칭 등에 사용)

# ===== 전처리(인코딩/스케일링) =====
from sklearn.preprocessing import LabelEncoder      # 카테고리를 숫자로 변환 (순서형)
from sklearn.preprocessing import MinMaxScaler      # 0~1 범위로 스케일링
from sklearn.preprocessing import StandardScaler    # 평균0, 표준편차1로 스케일링
from sklearn.preprocessing import RobustScaler      # 이상치에 강한 스케일링 (median/IQR 기반)

# ===== 모델 =====
from sklearn.linear_model import LinearRegression       # 선형 회귀 모델
from sklearn.linear_model import LogisticRegression     # 로지스틱 회귀(분류) 모델
from sklearn.tree import DecisionTreeClassifier         # 의사결정나무 분류 모델
from sklearn.dummy import DummyClassifier               # 베이스라인(단순 규칙) 분류 모델
from sklearn.ensemble import RandomForestRegressor      # 랜덤포레스트 회귀 모델
from sklearn.ensemble import RandomForestClassifier     # 랜덤포레스트 분류 모델(여러 트리의 앙상블)
from xgboost import XGBClassifier                       # XGBoost 분류 모델(부스팅 기반)

# ===== 파이프라인 =====
from sklearn.pipeline import make_pipeline              # 전처리+모델 여러 단계를 하나로 묶기

# ===== 데이터 분리 / 교차검증 =====
from sklearn.model_selection import train_test_split    # 데이터를 학습용/테스트용으로 분리
from sklearn.model_selection import KFold               # 데이터를 K개로 분할(교차검증용, 회귀에 적합)
from sklearn.model_selection import StratifiedKFold     # 클래스 비율을 유지하며 K개로 분할(불균형 분류용)
from sklearn.model_selection import cross_val_score     # 교차검증으로 단일 지표 반복 평가
from sklearn.model_selection import cross_validate      # 교차검증으로 여러 지표·학습점수까지 한번에 평가
from sklearn.model_selection import learning_curve      # 표본 크기에 따른 학습/검증 성능 변화 확인

# ===== 평가지표 =====
from sklearn.metrics import accuracy_score              # 분류 모델의 정확도 계산
from sklearn.metrics import confusion_matrix            # 실제/예측 결과를 교차표로 정리(TP/FP/TN/FN)
from sklearn.metrics import classification_report       # 정밀도·재현율·F1 등을 한번에 요약
from sklearn.metrics import precision_score             # 정밀도(예측한 것 중 실제로 맞은 비율)
from sklearn.metrics import recall_score                # 재현율(실제 정답 중 모델이 잡아낸 비율)
from sklearn.metrics import f1_score                     # 정밀도와 재현율의 조화평균
from sklearn.metrics import roc_curve                    # ROC 곡선 좌표 계산
from sklearn.metrics import roc_auc_score                # ROC 곡선 아래 면적(AUC)
from sklearn.metrics import precision_recall_curve       # 정밀도-재현율 곡선 좌표 계산
from sklearn.metrics import average_precision_score      # PR 곡선 아래 면적(AP)
from sklearn.metrics import log_loss                      # 분류 확률 예측의 손실(로그 손실)

In [3]:
import matplotlib.pyplot as plt
import urllib.request, zipfile

RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)

DATA_DIR = Path("data")
DATA_DIR.mkdir(exist_ok=True)

def fetch_uci(url, zip_name, member):
    """UCI 정적 저장소의 zip을 내려받아 data/에 풀고, CSV 경로를 돌려줍니다."""
    csv_path = DATA_DIR / member
    if not csv_path.exists():
        zip_path = DATA_DIR / zip_name
        if not zip_path.exists():
            print(f"내려받는 중… {zip_name}")
            urllib.request.urlretrieve(url, zip_path)
        with zipfile.ZipFile(zip_path) as z:
            z.extract(member, DATA_DIR)
    return csv_path

shoppers = pd.read_csv(fetch_uci(
    "https://archive.ics.uci.edu/static/public/468/online+shoppers+purchasing+intention+dataset.zip",
    "online_shoppers.zip", "online_shoppers_intention.csv"))

NUM_COLS = ["Administrative", "Administrative_Duration", "Informational",
            "Informational_Duration", "ProductRelated", "ProductRelated_Duration",
            "BounceRates", "ExitRates", "PageValues", "SpecialDay"]

X = shoppers[NUM_COLS]
y = shoppers["Revenue"].astype(int)

print(f"쇼핑 세션 데이터: {shoppers.shape[0]:,}행 × {shoppers.shape[1]}열")
print(f"구매 전환율(양성 비율): {y.mean():.4f}")
print("\n→ 준비 완료. 이제 여러분 차례입니다.")

내려받는 중… online_shoppers.zip
쇼핑 세션 데이터: 12,330행 × 18열
구매 전환율(양성 비율): 0.1547

→ 준비 완료. 이제 여러분 차례입니다.


In [ ]:
# [C2] 문제 1. 혼동행렬로 0.89를 쪼갠다
# ⌨️ 문제 1 — 정확도 하나를 네 칸으로 쪼개기
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (confusion_matrix, accuracy_score, precision_score,
                             recall_score, f1_score)

# 여기에 코드를 작성하세요 (분리 → 학습 → 혼동행렬 → 지표 4종 → 놓친 사람 수)

#[문제 1]
#1) 지난 순서에서 채택한 모델을 학습합니다.
#   RandomForestClassifier(n_estimators=300, min_samples_leaf=20, random_state=42)
#2) 테스트셋에 대한 혼동행렬을 구해 TN · FP · FN · TP를 각각 출력합니다.
#3) 정확도 · 정밀도 · 재현율 · F1을 계산해 함께 출력합니다.
#4) "실제 구매 세션 중 몇 건을 놓쳤는가"를 건수와 비율로 설명합니다.

X = shoppers[NUM_COLS]
y = shoppers['Revenue'].astype(int)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size = 0.2, stratify= y, random_state= 42)

rf = RandomForestClassifier(n_estimators = 300, min_samples_leaf = 20, random_state = 42)
rf.fit(X_train, y_train)

y_pred = rf.predict(X_test)

cm = confusion_matrix(y_test, y_pred)

tn, fp, fn, tp = cm.ravel()

print("2) 테스트셋에 대한 혼동행렬을 구해 TN · FP · FN · TP를 각각 출력합니다.")
print("\n혼동행렬 값")
print(f"tn: {tn}")
print(f"fp: {fp}")
print(f"fn: {fn}")
print(f"tp: {tp}")





#3) 정확도 · 정밀도 · 재현율 · F1을 계산해 함께 출력합니다.

혼동행렬 값
tn: 2002
fp: 82
fn: 186
tp: 196


In [42]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
average_precision = average_precision_score(y_test, y_pred)

print("#3) 정확도 · 정밀도 · 재현율 · F1을 계산해 함께 출력합니다.")
print("#4) 실제 구매 세션 중 몇 건을 놓쳤는가를 건수와 비율로 설명합니다.")
print(f"\naccuracy: {accuracy:.2f}%")
print(f"precision: {precision:.2f}%")
print(f"recall: {recall:.2f}%")
print(f"f1: {f1:.2f}%")
print(f"average_precision: {average_precision:.2f}%")
print(f"\n실제 구매 건수 중 놓친 건수 및 비율: 총{fn + tp}건 중 {fn}건 놓침, {100 - (recall * 100):.2f}%")

#3) 정확도 · 정밀도 · 재현율 · F1을 계산해 함께 출력합니다.
#4) 실제 구매 세션 중 몇 건을 놓쳤는가를 건수와 비율로 설명합니다.

accuracy: 0.88%
precision: 0.88%
recall: 0.26%
f1: 0.40%
average_precision: 0.34%

실제 구매 건수 중 놓친 건수 및 비율: 총382건 중 115건 놓침, 74.35%


In [31]:
# [C3] 문제 2. 지표 5종을 교차 검증으로 한 번에 잰다
# ⌨️ 문제 2 — cross_validate로 다섯 지표를 동시에
from sklearn.model_selection import StratifiedKFold, cross_validate
from sklearn.metrics import roc_auc_score

# 여기에 코드를 작성하세요 (5겹 CV로 지표 5종 → 평균±표준편차 → 기준선 대비)

#[문제 2]
#1) cross_validate로 다섯 지표를 한 번에 구합니다.
#   accuracy · precision · recall · f1 · average_precision(AP)
#2) 각각 평균 ± 표준편차로 출력합니다.
#3) AP를 양성 비율과 비교하고, 참고로 ROC-AUC도 구해 함께 확인합니다.
#4) 다섯 지표 중 표준편차가 가장 큰 것이 무엇인지, 왜 그런지 설명합니다.

skf = StratifiedKFold(n_splits = 5, shuffle = True, random_state = 42)

scoring = ["accuracy", "precision", "recall", "f1", "average_precision", "roc_auc"]

results = cross_validate(rf, X_train, y_train, cv = skf, scoring = scoring)

positive_rate = y_train.mean()

print(f"양성 비율(무작위 AP 기준선): {positive_rate:.4f}")

summary = []

for metric in scoring:
    scores = results[f"test_{metric}"]
    summary.append({"지표": metric, "평균": scores.mean(), "표준편차": scores.std()})

summary_df = pd.DataFrame(summary)
print(summary_df.round(4).to_string(index=False))

ap_row = summary_df[summary_df["지표"] == "average_precision"].iloc[0]
print(f"AP({ap_row['평균']:.4f}) vs 양성비율({positive_rate:.4f}) → 차이: {ap_row['평균']-positive_rate:.4f}")

# 4) 표준편차가 가장 큰 지표 찾기
worst_idx = summary_df["표준편차"].idxmax()
print(f"\n표준편차가 가장 큰 지표: {summary_df.loc[worst_idx, '지표']} (표준편차={summary_df.loc[worst_idx, '표준편차']:.4f})")

양성 비율(무작위 AP 기준선): 0.1547
               지표     평균   표준편차
         accuracy 0.8984 0.0062
        precision 0.7285 0.0320
           recall 0.5491 0.0282
               f1 0.6257 0.0237
average_precision 0.7289 0.0208
          roc_auc 0.9111 0.0079
AP(0.7289) vs 양성비율(0.1547) → 차이: 0.5742

표준편차가 가장 큰 지표: precision (표준편차=0.0320)


In [71]:
# [C4] 문제 3. 임계값을 훑어 운영점을 정한다
# ⌨️ 문제 3 — 용량 제약(쿠폰 500장)을 임계값으로 번역하기

# 여기에 코드를 작성하세요 (임계값 표 → F1 최적 → 상위 500건 역산 → 성과 보고)

#[문제 3]
#1) 임계값을 0.10부터 0.70까지 훑으며 양성 예측 건수 · 정밀도 · 재현율 · F1을 표로 만듭니다.
#2) F1이 가장 높은 임계값을 찾습니다.
#3) "쿠폰 500장" 제약을 만족하는 임계값을 역산합니다.
#   (확률 상위 500건만 고르려면 임계값이 얼마여야 하는가)
#4) 그 운영점에 실제 구매 세션이 몇 건 포함되는지 계산해 마케팅팀에 보고합니다.
proba = rf.predict_proba(X_test)[:, 1]


thresholds = np.sort(np.unique(proba))[::-1]
results = []

for th in thresholds:
    y_pred = (proba >= th).astype(int)
    n_positive = y_pred.sum()
    prec = precision_score(y_test, y_pred, zero_division=0)
    rec = recall_score(y_test, y_pred, zero_division=0)
    f1 = f1_score(y_test, y_pred, zero_division= 0)
    results.append({
        "임계값": th,
        "양성예측건수": n_positive,
        "정밀도": prec,
        "재현율": rec,
        "F1": f1
    })

result_df = pd.DataFrame(results)
print("=== 1) 임계값별 표 ===")
print(result_df[result_df["양성예측건수"] == 500].to_string(index=False))

=== 1) 임계값별 표 ===
     임계값  양성예측건수   정밀도      재현율       F1
0.268296     500 0.568 0.743455 0.643991


In [40]:
# [C5] 문제 4. `class_weight` 보정, 채택할 것인가
# ⌨️ 문제 4 — 보정 전후를 같은 방식으로 비교

# 여기에 코드를 작성하세요 (balanced CV 지표 → 기본과 비교표 → 혼동행렬 대조)

#[문제 4]
#1) class_weight="balanced"를 준 같은 모델로 교차 검증 지표 5종을 다시 구합니다.
#2) 문제 2의 기본 설정과 나란히 표로 비교합니다.
#3) 홀드아웃 혼동행렬도 함께 구해 TP·FN이 어떻게 달라졌는지 확인합니다.
#4) 채택할 것인지 결정하고, 그 이유를 적습니다.
#   (임계값 조정으로도 같은 효과를 낼 수 있다는 점을 함께 고려합니다)

rf_balanced = RandomForestClassifier(n_estimators=300, class_weight="balanced", random_state=42)
results_balanced = cross_validate(rf_balanced, X_train, y_train, cv = skf, scoring=scoring)

rf_default = RandomForestClassifier(n_estimators=300, random_state=42)
results_default = cross_validate(rf_default, X_train, y_train, cv = skf, scoring=scoring)

compare = []
for metric in scoring:
    default_mean = results_default[f"test_{metric}"].mean()
    balanced_mean = results_balanced[f"test_{metric}"].mean()
    compare.append({
        "지표": metric,
        "기본": default_mean,
        "balanced": balanced_mean,
        "차이": balanced_mean - default_mean
    })

compare_df = pd.DataFrame(compare)
print("=== 기본 vs balanced 비교표 ===")
print(compare_df.round(4).to_string(index=False))

# ===== 3) 홀드아웃 혼동행렬 비교 =====
rf_default.fit(X_train, y_train)
rf_balanced.fit(X_train, y_train)

cm_default = confusion_matrix(y_test, rf_default.predict(X_test))
cm_balanced = confusion_matrix(y_test, rf_balanced.predict(X_test))

print("\n=== 기본 설정 혼동행렬 ===")
print(cm_default)
tn, fp, fn, tp = cm_default.ravel()
print(f"TP={tp}, FN={fn}")

print("\n=== class_weight=balanced 혼동행렬 ===")
print(cm_balanced)
tn, fp, fn, tp = cm_balanced.ravel()
print(f"TP={tp}, FN={fn}")


=== 기본 vs balanced 비교표 ===
               지표     기본  balanced      차이
         accuracy 0.8970    0.8887 -0.0083
        precision 0.7126    0.6179 -0.0947
           recall 0.5636    0.7418  0.1782
               f1 0.6288    0.6737  0.0449
average_precision 0.7186    0.7107 -0.0080
          roc_auc 0.9036    0.9031 -0.0005

=== 기본 설정 혼동행렬 ===
[[2004   80]
 [ 177  205]]
TP=205, FN=177

=== class_weight=balanced 혼동행렬 ===
[[1902  182]
 [ 115  267]]
TP=267, FN=115


## 모델 카드 v4 — 구매 전환 예측

- 데이터: UCI Online Shoppers (12,330 세션, 수치형 10개 열, 양성 15.5%)
- 문제 유형: 분류 (불균형)
- 검증 방식: StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
- 모델: RandomForest(n_estimators=300, min_samples_leaf=20)
- **주 지표와 선택 근거: { 무엇을 왜 — 정확도만으로 판단하지 않는 이유를 포함 }**
- **성능(CV, 평균 ± 표준편차)**
  - accuracy           0.8984 0.0062
  - precision          0.7285 0.0320
  - recall             0.5491 0.0282
  - f1                 0.6257 0.0237
  - average_precision  0.7289 0.0208
  
  - **양성 비율 {0.1547}와 AP {0.7289} 비교 → 차이: 0.5742**
  
  - **혼동행렬(임계값 0.5):**
- tn: 2002
- fp: 82
- fn: 186
- tp: 196
  - **놓친 구매 세션  총382건 중 115건 놓침, 74.35%)  ← 정확도만 봤을 때 보이지 않던 숫자**

- **운영점: 임계값 {0.26} (용량 제약 {503}건)**
  - **그 지점의 정밀도 { } / 재현율 { } → 실제 구매 세션 { }건 포함**
  - **선택 근거: { F1 최적점이 아니라 예산 제약인 이유 }**
  - 임계값: 0.268296
  - 양성예측건수: 500     
  - precision: 0.568
  - recall: 0.743455
  - F1: 0.643991
  - 쿠폰 500개로 한정되어 있다보니 양성예측건수가 500이되는 임계값 0.268296으로 정했습니다.
  
- **class_weight 보정: 채택 { 여부 } — 근거: { AP 변화와 임계값 대안을 함께 언급 }**
- 한계 & 다음 단계: { 개입 효과와 구매 확률의 차이 / 범주형 8개 열 미사용 / 운영점은 검증·OOF 예측에서 다시 선택 등 }
- AI 사용 내역: { 무엇을 물었나 / 무엇을 검증했나 }